File: factory/Factory.java

public class Factory {

    /* Axe Class */
    private interface Axe {
        String getMaterial();
    }

    private enum Material {
        WOOD,
        STONE,
        IRON
    }

    // Wooden Axe
    private class WoodenAxe implements Axe {
        @Override
        public String getMaterial() {
            return "wood";
        }
    }

    // Stone Axe
    private class StoneAxe implements Axe {
        @Override
        public String getMaterial() {
            return "stone";
        }
    }

    // Iron Axe
    private class IronAxe implements Axe {
        @Override
        public String getMaterial() {
            return "iron";
        }
    }

    /* Simple factory*/
    public abstract class AxeFactory {

        public abstract Axe createAxe(Material material);
    }

    public class WoodAxeFactory extends AxeFactory {
        @Override
        public Axe createAxe(Material material) {
            if(material != Material.WOOD)
                throw new IllegalArgumentException("Invalid material");
            return new WoodenAxe();
        }
    }

    public class StoneAxeFactory extends AxeFactory {
        @Override
        public Axe createAxe(Material material) {
            if(material != Material.STONE)
                throw new IllegalArgumentException("Invalid material");
            return new StoneAxe();
        }
    }

    public class IronAxeFactory extends AxeFactory {
        @Override
        public Axe createAxe(Material material) {
            if(material != Material.IRON)
                throw new IllegalArgumentException("Invalid material");
            return new IronAxe();
        }
    }

    public void testAxeFactory() {
        AxeFactory factory;

        factory = new WoodAxeFactory();
        var woodAxe = factory.createAxe(Material.WOOD);

        factory = new StoneAxeFactory();
        var stoneAxe = factory.createAxe(Material.STONE);

        factory = new IronAxeFactory();
        var ironAxe = factory.createAxe(Material.IRON);

        System.out.println("WoodAxe is made of " + woodAxe.getMaterial());
        System.out.println("StoneAxe is made of " + stoneAxe.getMaterial());
        System.out.println("IronAxe is made of " + ironAxe.getMaterial());
    }

    public static void main(String[] args) {
        Factory testFactory = new Factory();
        testFactory.testAxeFactory();
    }
}

File: factory/Factory.py
from enum import Enum
# python 3.8 and above
from typing import Protocol, Dict

class Material(Enum):
    WOOD = 1
    STONE = 2
    IRON = 3

# protocol is a new feature in python 3.8, it is similar to interface in java
# Note that it is only for type checking, it does not enforce the implementation
# https://docs.python.org/3/library/typing.html#typing.Protocol

class Axe(Protocol):
    def getMaterial(self):
        pass

class WoodAxe(Axe):
    def getMaterial(self):
        return "wood"

class StoneAxe(Axe):
    def getMaterial(self):
        return "stone"

class IronAxe(Axe):
    def getMaterial(self):
        return "iron"

class AxeFactory(Protocol):
    def createAxe(self, material: Material):
        pass

class WoodAxeFactory(AxeFactory):
    def createAxe(self, material: Material):
        if material == Material.WOOD:
            return WoodAxe()
        return None

class StoneAxeFactory(AxeFactory):
    def createAxe(self, material: Material):
        if material == Material.STONE:
            return StoneAxe()
        return None
    
class IronAxeFactory(AxeFactory):
    def createAxe(self, material: Material):
        if material == Material.IRON:
            return IronAxe()
        return None

def testAxeFactory():
    factory: AxeFactory = WoodAxeFactory()
    woodAxe = factory.createAxe(Material.WOOD)

    factory: AxeFactory = StoneAxeFactory()
    stoneAxe = factory.createAxe(Material.STONE)

    factory: AxeFactory = IronAxeFactory()
    ironAxe = factory.createAxe(Material.IRON)

    print("WoodAxe is made of " + woodAxe.getMaterial())
    print("StoneAxe is made of " + stoneAxe.getMaterial())
    print("IronAxe is made of " + ironAxe.getMaterial())

if __name__ == "__main__":
    testAxeFactory()
